cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeEngineCredentials.java

```diff
@@ -0,0 +1,142 @@
+import com.google.gson.Gson;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Date;
+
+/**
+ * Fecth and refresh token
+ */
+public final class ComputeEngineCredentials {
+  private static final Logger LOG = LoggerFactory.getLogger(ComputeEngineCredentials.class);
+  private static final Gson GSON = new Gson();
+
+  /**
+   * Time (in millisecond) to refresh the credentials before it expires.
+   */
+  private static final long REFRESH_BEFORE_EXPIRATION_TIME = 1000 * 60 * 10;
+  private static final int NUMBER_OF_RETRIES = 10;
+  private static final int MIN_WAIT_TIME = 5000;
+  private static final int MAX_WAIT_TIME = 10000;
+
+  private static GoogleCredentials cachedGoogleCredentials = null;
+
+  private ComputeEngineCredentials() {
+
+  }
+
+  /**
+   * Fetches and returns GoogleCredentials locally if endPoint is null, or remotely from the provided endPoint.
+   *
+   * @param endPoint endpoint to be used for fetching GoogleCredentials from.
+   * @return GoogleCredentials
+   * @throws IOException
+   */
+  public static GoogleCredentials getComputeEngineCredentials(String endPoint) throws IOException {
+    if (cachedGoogleCredentials != null && System.currentTimeMillis() <
+      cachedGoogleCredentials.getAccessToken().getExpirationTime().getTime() - REFRESH_BEFORE_EXPIRATION_TIME) {
+      // Cached GoogleCredentials is still valid.
+      return cachedGoogleCredentials;
+    }
+
+    if (endPoint != null) {
+      return getRemoteCredentials(endPoint);
+    }
+
+    return getLocalCredentials();
+  }
+
+  private static GoogleCredentials getLocalCredentials() throws IOException {
+    try {
+      cachedGoogleCredentials = com.google.auth.oauth2.ComputeEngineCredentials.create();
+      cachedGoogleCredentials.refreshAccessToken();
+      return cachedGoogleCredentials;
+    } catch (IOException e) {
+      throw new IOException("Unable to get credentials from the environment. "
+                              + "Please explicitly set the account key.", e);
+    }
+  }
+
+  private static GoogleCredentials getRemoteCredentials(String endPoint) throws IOException {
+    ExponentialBackOff backOff = new ExponentialBackOff.Builder()
+      .setInitialIntervalMillis(MIN_WAIT_TIME)
+      .setMaxIntervalMillis(MAX_WAIT_TIME).build();
+
+    for (int i = 0; i < NUMBER_OF_RETRIES; i++) {
+      try {
+        cachedGoogleCredentials = fetchCredentials(endPoint);
+        return cachedGoogleCredentials;
+      } catch (Exception ex) {
+        LOG.warn("Failed to fetch GoogleCredentials. Will retry.", ex);
+      }
+      try {
+        Thread.sleep(backOff.nextBackOffMillis());
+      } catch (InterruptedException e) {
+        // No-op
```

Comment:
Don't do a no-op on interruptedException. It can block shutdown in some case. It's better to log and return